Microsoft.Toolkit.Uwp.UI.Controls.Layout/Carousel/CarouselItemAutomationPeer.cs

```diff
@@ -0,0 +1,211 @@
+        }
+
+        /// <summary>Removes the current element from the collection of selected items.</summary>
+        public void RemoveFromSelection()
+        {
+            // Cannot remove the selection of a Carousel control.
+        }
+
+        /// <summary>Clears any existing selection and then selects the current element.</summary>
+        public void Select()
+        {
+            CarouselItem owner = this.OwnerCarouselItem;
+            Carousel parent = owner.ParentCarousel;
+            parent?.SetSelectedItem(owner);
+        }
+
+        /// <summary>
+        /// Gets the control type for the element that is associated with the UI Automation peer.
+        /// </summary>
+        /// <returns>The control type.</returns>
+        protected override AutomationControlType GetAutomationControlTypeCore()
+        {
+            return AutomationControlType.ListItem;
+        }
+
+        /// <summary>
+        /// Called by GetClassName that gets a human readable name that, in addition to AutomationControlType,
+        /// differentiates the control represented by this AutomationPeer.
+        /// </summary>
+        /// <returns>The string that contains the name.</returns>
+        protected override string GetClassNameCore()
+        {
+            return Owner.GetType().Name;
+        }
+
+        /// <summary>
+        /// Called by GetName.
+        /// </summary>
+        /// <returns>
+        /// Returns the first of these that is not null or empty:
+        /// - Value returned by the base implementation
+        /// - Name of the owning CarouselItem
+        /// - Carousel class name
+        /// </returns>
+        protected override string GetNameCore()
+        {
+            string name = AutomationProperties.GetName(this.OwnerCarouselItem);
+            if (!string.IsNullOrEmpty(name))
+            {
+                return name;
+            }
+
+            name = this.OwnerCarouselItem.Name;
+            if (!string.IsNullOrEmpty(name))
+            {
+                return name;
+            }
+
+            var textBlock = this.OwnerCarouselItem.FindDescendant<TextBlock>();
+            if (textBlock != null)
+            {
+                return textBlock.Text;
+            }
+
+            return base.GetNameCore();
+        }
+
+        /// <summary>
+        /// Called by GetAutomationId that gets the **AutomationId** of the element that is associated with the automation peer.
+        /// </summary>
+        /// <returns>
+        /// The string that contains the automation ID.
+        /// </returns>
+        protected override string GetAutomationIdCore()
+        {
+            var automationId = base.GetAutomationIdCore();
+            if (!string.IsNullOrEmpty(automationId))
+            {
+                return automationId;
+            }
+
+            if (this.OwnerCarouselItem != null)
+            {
+                return this.GetNameCore();
+            }
```

Comment:
Is it best practice to return the name if there is no automation id? I think default is just return nothing instead of using the name...

Reply:
The name allows you to differentiate them if the automation id is not set. This is a pattern I've seen applied in other control libraries also with parent/child UI elements.

As names are unique, it also provides that additional layer for locating items in the visual tree.

Reply:
WinUI and the default UWP controls don't seem to do this though, e.g. NumberBox, TabView or Button all just have a blank AutomationID if you don't set that value. I would argue that the WCT should align with the default behavior of the platform here. Do you have examples at hand that follow the proposed pattern at hand? Would love to know them.

Reply:
To hand, one parent/child UI element relationship which has this implemented is the Telerik UI for UWP LoopingListItem.

- [LoopingListItemAutomationPeer](https://github.com/telerik/UI-For-UWP/blob/3f9c86ba361cbc8960be9493c6eb2b91092e355b/Controls/Primitives/Primitives.UWP/LoopingList/AutomationPeers/LoopingListItemAutomationPeer.cs#L115)

Reply:
Interestingly, WPF's `FrameworkElement` implementation does something similar where it gets the name of the underlining control. 

- [WPF FrameworkElementAutomationPeer](https://github.com/dotnet/wpf/blob/d49f8ddb889b5717437d03caa04d7c56819c16aa/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Automation/Peers/FrameworkElementAutomationPeer.cs#L30)

I wonder if UWP follows suit and is implemented in a similar way. Although this isn't quite the `GetNameCore` which could provide even more detail.

Reply:
I did also provide this same implementation for the [`BladeItemAutomationPeer`](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/Microsoft.Toolkit.Uwp.UI.Controls.Layout/BladeView/BladeItemAutomationPeer.cs) in the toolkit too.

Reply:
The WPF peer uses the equivalent of x:Name, not the UIA name (GetNameCore) so matching that behavior makes more sense. From my sloppy investigation, I found out that the UIA name get's ignored for UIA ID, the Name property is the important thing for the UIA name, so calling GetNameCore is not the right approach here.

Reply:
Good shout. I've had a play around and this does work how I thought it did if not there, so I've removed this from this control and my implementation for the `BladeItem` automation peer also.

Reply:
Awesome, thanks for changing that!